### PR TITLE
Introduce EnumDesc, add ability to output source location attribute

### PR DESCRIPTION
**Pull request by andrew-boyarshin**
Not currently exposed, but can easily be plumbed as opt-in command line option. Note, that not all Desc items have Location attribute set, since for some it doesn't make any sense.

### CODE REVIEW

File: sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs [hunk 1]
```diff
@@ -268,6 +281,27 @@ public void BeginFunctionOrDelegate<TCustomAttrGeneratorData>(in FunctionOrDeleg
             }
         }
 
+        private void WriteSourceLocation(CXSourceLocation location, bool inline)
```
Comment (tannergooding) [inline sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs:284]: I'm concerned this is going to end up very noisy across regenerations. The anonymous struct names/remappings is already a pain point for some users

**Comment by tannergooding** [comment]
The rest of the changes look reasonable to me.

File: sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs (hunk 1 above)
Comment (andrew-boyarshin) [reply sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs:284]: It's opt-in. It is useful for diagnostics (aka "where did this come from?"), but not so much for end-user of the bindings. I don't expect SharpGen-generated bindings or e.g. TerraFX.Interop to contain this info. But e.g. win32metadata and SharpGen might use this information to enrich logging or other forms of diagnostics.

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs:284]: Right, I understand the benefit.

I was just iterating the concern around noise in practice. Many diffs would become nearly unreadable due to minor changes in information such as whitespace addition.

If it's being added here, a corresponding switch and basic test validating it works and doesn't regress should also be added.

Comment (andrew-boyarshin) [reply sources/ClangSharp.PInvokeGenerator/CSharp/CSharpOutputBuilder.VisitDecl.cs:284]: Done. I've added 2 tests that cover all important cases.

File: sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfigurationOptions.cs [hunk 2]
```diff
@@ -58,5 +58,7 @@ public enum PInvokeGeneratorConfigurationOptions
         GeneratePreviewCode = 0x00800000,
 
         GenerateTemplateBindings = 0x01000000,
+
+        GenerateSourceLocationAttribute = 1 << 25,
```
Comment (tannergooding) [inline sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfigurationOptions.cs:62]: nit: This should be either `0x02000000` or all entries should be updated to use the `1 << n` syntax.

I don't have a particular preference for which is done, I'd just like for it to be consistent

**Comment by tannergooding** [comment]
Changes LGTM, minus the differing flag value

File: sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfigurationOptions.cs (hunk 2 above)
Comment (andrew-boyarshin) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfigurationOptions.cs:62]: I'll update the rest to use bit shifting to be explicit and simplify visual validation of flag values.

Comment (andrew-boyarshin) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfigurationOptions.cs:62]: Done.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**